code/game/machinery/crusher_piston.dm

```diff
@@ -0,0 +1,328 @@
+	//Move all the items in the move list
+	for(var/atom/movable/AM in items_to_move)
+		if(!AM.simulated)
+			items_to_move -= AM
+			continue
+		if(istype(AM,/obj/machinery/crusher_piston))
+			items_to_move -= AM
+			continue
+		items_to_move -= AM
+		//If a item could not be moved. Add it to the crush list
+		if(!AM.piston_move())
+			items_to_crush += AM
+		CHECK_TICK
+	
+	//Destroy all the items in the crush list
+	for(var/atom/movable/AM in items_to_crush)
+		items_to_crush -= AM
+		AM.crush_act()
+		CHECK_TICK
+
+	process_lock = 0
+
+
+/obj/machinery/crusher_base/proc/crush_start()
+	action = "extend"
+
+
+/obj/machinery/crusher_base/proc/crush_abort()
+	//Abort the crush
+	//Retract all the pistons, ...
+	action = "retract"
+	//TODO-Crusher: Add more stages to the piston so retracting will not cause a visual bug
+
+
+//Piston Stage 1
+/obj/machinery/crusher_piston
+	name = "Trash compactor piston"
+	desc = "A colossal piston used for crushing garbage."
+	icon = 'icons/obj/machines/crusherpiston.dmi' //Placeholder TODO: Get a proper icon
+	icon_state = "piston_0" 
+	density = 0
+	anchored = 1
+	pixel_y = -64
+	var/stage = 0 //The stage of the piston
+	var/obj/machinery/crusher_base/crs_base //Crusher Base the piston is linked to
+	var/obj/effect/piston_blocker/pb1
+	var/obj/effect/piston_blocker/pb2
+	var/obj/effect/piston_blocker/pb3
+
+/obj/machinery/crusher_piston/proc/extend_0_1()
+	icon_state="piston_0_1"
+	stage = 1
+	pb1 = new(locate(x,y,z))
+	for(var/atom/movable/AM in get_turf(locate(x,y,z)))
+		crs_base.items_to_move += AM
+
+/obj/machinery/crusher_piston/proc/extend_1_2()
+	icon_state="piston_1_2"
+	stage = 2
+	pb2 = new(locate(x,y-1,z))
+	for(var/atom/movable/AM in get_turf(locate(x,y-1,z)))
+		crs_base.items_to_move += AM
+
+/obj/machinery/crusher_piston/proc/extend_2_3()
+	icon_state="piston_2_3"
+	stage = 3
+	pb3 = new(locate(x,y-2,z))
+	for(var/atom/movable/AM in get_turf(locate(x,y-2,z)))
+		crs_base.items_to_move += AM
+
+/obj/machinery/crusher_piston/proc/retract_3_0()
+	icon_state="piston_3_0"
+	stage = 0
+	qdel(pb1)
+	qdel(pb2)
+	qdel(pb3)
+
+/obj/effect/piston_blocker
+	name = "Trash compactor piston blocker" //TODO: Fix Name
+	desc = "A colossal piston used for crushing garbage."
+	density = 1
+	anchored = 1
+
+//
+// The piston_move proc for various objects
+//
+// Return 1 if it could successfully be moved
+// Return 0 if it could not be moved
+/atom/movable/proc/piston_move()
+	var/turf/T = get_turf(src)
+
+	var/list/valid_turfs = list()
+	for(var/dir_to_test in cardinal)
+		var/turf/new_turf = get_step(T, dir_to_test)
+		if(!new_turf.contains_dense_objects())
+			valid_turfs |= new_turf
+
+	while(valid_turfs.len)
+		T = pick(valid_turfs)
+		valid_turfs -= T
+
+		if(src.forceMove(T))
+			return 1
+	// We failed to move our target
+	return 0
+
+/mob/living/piston_move(var/crush_damage)
+	for(var/i = 1, i <= PISTON_MOVE_DIVISOR, i++)
+		adjustBruteLoss(round(PISTON_MOVE_DAMAGE / PISTON_MOVE_DIVISOR))
+	SetStunned(5)
```

Comment:
This should be `Stun(5)` and `Weaken(5)`, unless you want to potentially reduce these numbers.